### ISSUE TITLE
HLS problems in iOS 

### ISSUE THREAD

**Issue by abdualelahM** [jhomlala/betterplayer#475]
Hi, jhomlala

I am very thankful for your great job. 

I am facing serval issue on iOS
- Progress bar is not responsive and not accurate. For example, when I seek backward the bar and the duration goes to 00:00 then it goes back to the correct duration.

- When I seek forward in a long video then I seek back, an error occurs and the player asks to retry (Sometimes more than once ).  I notice this issue occurs when I seek to none buffered duration then I seek backward.

- Background audio overlaps with other sounds in the device.

Kind regards,

**Comment by m1n093** [comment]
I'm having the same issue.

**Comment by abdualelahM** [comment]
The issue, in short, the listener returning incorrect values (Old values) so when seeking backward or forward multiple times the behavior goes wrong and the result is incorrect seeking position. 

My Workaround is :
- implemented a method channel for seeking + - 10 seconds in ios av player and not relying on the duration listener. 
- Implemented a costume progress bar that does update on drag update, only on drag end. (the behavior is smilier to the Youtube player app).    

I hope that helps @m1n093 @jhomlala 

Any further clarification I am happy to assist.

**Comment by anhnt** [comment]
@abdualelahM Thanks! Did you mean your code above fix prob on real ios device?

**Comment by abdualelahM** [comment]
@anhnt 

I was convinced that my code worked until my build reached TestFlight. The issue definitely is from position listner either from native code part or from dart part, it could be both. However, I made a position correction trick. It is not the best but it is working for me. 

I hope it helps. 

```dart
  bool forwardingMode = false;
  bool seekForward = false;
  bool seekBackward = false;

  Duration seekingPos = Duration(microseconds: 0);
  Duration lastPos = Duration(microseconds: 0);

  @override
  Widget build(BuildContext context) {
    Duration seekToRelativePosition(Offset globalPosition) {
      final RenderObject? renderObject = context.findRenderObject();
      if (renderObject != null) {
        final box = renderObject as RenderBox;
        final Offset tapPos = box.globalToLocal(globalPosition);
        final double relative = tapPos.dx / box.size.width;
        if (relative > 0) {
          return controller!.value.duration! * relative;
        }
      }
      return Duration(microseconds: 0);
    }

    final bool enableProgressBarDrag = betterPlayerController!
        .betterPlayerConfiguration.controlsConfiguration.enableProgressBarDrag;


    if (forwardingMode) {
      if (seekForward) {
        print("Forward Correction");
        currentPostion = seekingPos;
        if (controller!.value.position.inMilliseconds >
            seekingPos.inMilliseconds) {
          forwardingMode = false;
          seekForward = false;
        }
      }
      if (seekBackward) {
        print("Backward Correction");

        currentPostion = seekingPos;
        if (controller!.value.position.inMilliseconds >
                seekingPos.inMilliseconds &&
            !(controller!.value.position.inMilliseconds <=
                    lastPos.inMilliseconds + 300 &&
                controller!.value.position.inMilliseconds >=
                    lastPos.inMilliseconds - 300)) {
          forwardingMode = false;
          seekBackward = false;
        }
      }
    } else {
      currentPostion = controller!.value.position;
    }
    return GestureDetector(
      onHorizontalDragStart: (DragStartDetails details) {
        if (!controller!.value.initialized || !enableProgressBarDrag) {
          return;
        }
      },
      onHorizontalDragUpdate: (DragUpdateDetails details) {},
      onHorizontalDragEnd: (DragEndDetails details) {},
      onTapDown: (TapDownDetails details) {
        if (!controller!.value.initialized || !enableProgressBarDrag) {
          return;
        }
        Duration value = seekToRelativePosition(details.globalPosition);
        print("Tap down");
        forwardingMode = false;
        seekBackward = false;
        seekForward = false;

        print(value.inMilliseconds.toString());
        print(controller!.value.position.inMilliseconds);

        if (value.inMilliseconds > controller!.value.position.inMilliseconds) {
          forwardingMode = true;
          seekForward = true;
          seekingPos = value;
          betterPlayerController!.seekTo(value);
        }

        if (value.inMilliseconds < controller!.value.position.inMilliseconds) {
          forwardingMode = true;
          seekBackward = true;
          seekingPos = value;
          lastPos = controller!.value.position;
          betterPlayerController!.seekTo(value);
        }
      },
      child: Center(
        child: Container(
          height: MediaQuery.of(context).size.height / 2,
          width: MediaQuery.of(context).size.width,
          color: Colors.transparent,
          child: !controller!.value.initialized
              ? Container()
              // modified version of  https://pub.dev/packages/audio_video_progress_bar 
              : ProgressBar(
                  progress: currentPostion,
                  total: controller!.value.duration! == null
                      ? Duration(milliseconds: 0)
                      : controller!.value.duration!,
                  onSeek: (value) {
                    forwardingMode = false;
                    seekBackward = false;
                    seekForward = false;
                    if (value.inMilliseconds >
                        controller!.value.position.inMilliseconds) {
                      forwardingMode = true;
                      seekForward = true;
                      seekingPos = value;
                      betterPlayerController!.seekTo(value);
                    }

                    if (value.inMilliseconds <
                        controller!.value.position.inMilliseconds) {
                      forwardingMode = true;
                      seekBackward = true;
                      seekingPos = value;
                      lastPos = controller!.value.position;
                      betterPlayerController!.seekTo(value);
                    }
                  },
                  progressBarColor: AppColors.colorOrange,
                  baseBarColor: Color.fromRGBO(188, 196, 204, 1),
                  thumbColor: AppColors.colorOrange,
                  thumbGlowColor: AppColors.colorOrange.withOpacity(.4),
                  barHeight: 3.0,
                  thumbRadius: 7,
                  thumbGlowRadius: 15,
                  onDragStart: () {
                    if (!controller!.value.initialized ||
                        !enableProgressBarDrag) {
                      return;
                    }

                    if (widget.onDragStart != null) {
                      widget.onDragStart!();
                    }
                  },
                  onDragUpdate: () {
                    if (widget.onDragUpdate != null) {
                      widget.onDragUpdate!();
                    }
                  },
                  onDragEnd: () {
                    if (widget.onDragEnd != null) {
                      widget.onDragEnd!();
                    }
                  },
                  ##),
        ),
      ),
    );
  }
}
```

**Comment by jhomlala** [comment]
Progress bar jump issue will be fixed in next version.

**Comment by jhomlala** [comment]
Fixed in 0.0.67.
